### DISCUSSION TITLE
How does the DateRangePicker work?

### DISCUSSION THREAD

**Discussion by juliangums** [malzariey/filament-daterangepicker-filter#58] [General]
Looks like a great plugin @malzariey 
I went through the documentation and I get how it works as a filter, but I am struggling to understand how it works as a form field. What I mean is `DateRangeFilter::make('created_at'),` makes total sense as you are applying it on one column. But `DateRangePicker::make('created_at'),` doesn't as I would need to save the date range as a start_date and end_date in the database. How does this work?

**Comment by juliangums** [comment]
@devhoussam1998 I can see you're also using it. Any idea?

**Comment by juliangums** [comment]
@mohssineAboutaj You're also using it. How did you do it?

**Comment by mohssineAboutaj** [comment]
@juliangums i guess you forgot to change the `created_at` type to `array`

**Reply by juliangums** [reply]
@mohssineAboutaj I'm not sure if I can follow you. Did you make `created_at` an array in your database? I basically have a `start_date` and `end_date` in my database. 

I was hoping I could do something like this to pull the dates from the daterange element:
```
                        Hidden::make('start_date')
                            ->value(function (Get $get) {
                                return $get('daterange') ? explode(' - ', $get('daterange'))[0] : '';
                            }),
                        Hidden::make('end_date')
                            ->value(function (Get $get) {
                                return $get('daterange') ? explode(' - ', $get('daterange'))[1] : '';
                            }),
```
But value() is not a method. I'm not sure what the correct method is, if there is one. Or is there any other way?

**Reply by juliangums** [reply]
I cannot do this `DateRangePicker::make(['start_date', 'end_date'])` as make() expects a string

**Comment by mohssineAboutaj** [comment]
@juliangums If you want to use the `start_date` and `end_date` to store the range you don't have to use this plugin, just use the default `DatePicker` in the **filamentphp**
but if you want to store the range in one column/field you have to create/change a column type in the `Model` and `Migration` and make it `array`
in my case, I decide to use separate fields like you
![image](https://github.com/malzariey/filament-daterangepicker-filter/assets/30943833/8c6a2201-decc-4125-bc33-124d5c16f04a)
here is the code
![image](https://github.com/malzariey/filament-daterangepicker-filter/assets/30943833/1add8ea6-4459-4d05-b2dd-bb53add47943)

**Reply by juliangums** [reply]
@mohssineAboutaj I think on a database level it makes much more sense to have it in two columns but for the UX depending on the use case it can be nicer to have the Daterangepicker. I solved mine like this now:

```
Hidden::make('start_date')
                            ->dehydrateStateUsing(fn (Get $get) => $get('daterange') ? Carbon::createFromFormat('d/m/Y', explode(' - ', $get('daterange'))[0])->format('Y-m-d') : ''),
                        Hidden::make('end_date')
                            ->dehydrateStateUsing(fn (Get $get) => $get('daterange') ? Carbon::createFromFormat('d/m/Y', explode(' - ', $get('daterange'))[1])->format('Y-m-d') : ''),
```

**Comment by juliangums** [comment]
@malzariey you might want to add this to the Readme as without it, the Picker might not be that useful for most people.

**Comment by malzariey** [comment]
Great discussion, once I am free, I plan to support multiple columns and restructure the plugin a little bit.

**Reply by sng82** [reply]
Did this ever happen? 
I've just run into the same problem as @juliangums, assuming there must be a way to have DateRangePicker populate two columns in the underlying database. I'm surprised this doesn't come up regularly.
If the functionality hasn't been added, it would be nice if the documentation mentioned the limitation. This discussion seems to be the only place it is mentioned and it's not easy to find.

**Reply by juliangums** [reply]
Don't think so but depending on your exact scenario you can use the hidden fields as described above or mutate the form data as described in https://filamentphp.com/docs/4.x/actions/create#customizing-data-before-saving or https://filamentphp.com/docs/4.x/resources/creating-records#customizing-data-before-saving

something like this
```php
CreateAction::make()
    ...
    ->mutateDataUsing(function (array $data): array {
        $data['start_date'] = $data['daterange'] ? Carbon::createFromFormat('d/m/Y', explode(' - ', $data['daterange'])[0])->format('Y-m-d') : null;
        $data['end_date'] = $data['daterange'] ? Carbon::createFromFormat('d/m/Y', explode(' - ', $data['daterange'])[1])->format('Y-m-d') : null;
        unset($data['daterange']);

        return $data;
    })
```

or this

```php
protected function mutateFormDataBeforeCreate(array $data): array
{
        $data['start_date'] = $data['daterange'] ? Carbon::createFromFormat('d/m/Y', explode(' - ', $data['daterange'])[0])->format('Y-m-d') : null;
        $data['end_date'] = $data['daterange'] ? Carbon::createFromFormat('d/m/Y', explode(' - ', $data['daterange'])[1])->format('Y-m-d') : null;
        unset($data['daterange']);

        return $data;
}
```

**Reply by sng82** [reply]
Thanks for this, very helpful! 

Your example would look fantastic in the README. _wink wink, nudge nudge!_ @malzariey

**Comment by bikash-s-bhandari** [comment]
why DateRangePicker component is not setting start and end date in filed??

DateRangePicker::make('date_range')
                                ->label('Weekend Date')
                                ->firstDayOfWeek(0)
                                ->startDate($lastSaturday)
                                ->endDate($lastSaturday)
                                ->disableRanges()
                                ->disabledDates($disabledDates)
                                ->separator(' - ')
                                ->displayFormat('MM/DD/YYYY'),

**Reply by malzariey** [reply]
First check if `defaultToday()` is working or not. 

Then check you variable value with dd before initializing the picker

**Reply by bikash-s-bhandari** [reply]
@malzariey  [Malzariey\FilamentDaterangepickerFilter\Fields\DateRangePicker::defaultToday does not exist.

in my filter section, i am using DateRangePicker  but its showing default start and end date empty in filed
![Screenshot from 2024-04-24 11-57-52](https://github.com/malzariey/filament-daterangepicker-filter/assets/33630192/a7ab2cc0-16b1-4a29-871c-24a6f79ea247)

i want to set start and end date like this intially
 
![Screenshot from 2024-04-24 11-58-00](https://github.com/malzariey/filament-daterangepicker-filter/assets/33630192/66026e5b-8522-4a99-874b-1a58d85ccb48)

**Reply by malzariey** [reply]
What version are you using?

**Reply by bikash-s-bhandari** [reply]
i am using "malzariey/filament-daterangepicker-filter": "^2.5",

**Reply by bikash-s-bhandari** [reply]
@malzariey  my filament admin panel version is  "filament/filament": "^3.2",

**Reply by malzariey** [reply]
Why are you using `DateRangePicker` , use `DateRangeFilter` instead

and `->defaultToday()` is available in the filter

**Reply by bikash-s-bhandari** [reply]
@malzariey  i am using filter inside filament form, there are lot of other filters as well and filters are applied by clicking apply filter buton, my code
```
 Filter::make('dependent_filters')
                ->form([
                    Grid::make(4)
                        ->schema([
                            DateRangePicker::make('date_range')
                                ->label('Weekend Date')
                                ->firstDayOfWeek(0)
                                ->startDate($lastSaturday)
                                ->endDate($lastSaturday)
                                ->disableRanges()
                                ->disabledDates($disabledDates)
                                ->separator(' - ')
                                ->displayFormat('MM/DD/YYYY'),
                            Select::make('category')
                                ->label('Categories(s)')
                                ->options(
                                   Category::pluck('name', 'id')
                                )
                                ->multiple()
                                ->live(true),
```
if i use DateRangeFilter, it cant be use inside Filament form, and also it will not work when i click apply filter button

**Reply by malzariey** [reply]
Try the new version v2.6.0
